File: android/src/main/java/net/xpece/android/widget/XpNumberPicker.java
package net.xpece.android.widget;

import android.annotation.TargetApi;
import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.widget.NumberPicker;

import java.lang.reflect.Field;

/**
 * Created by Eugen on 06.05.2016.
 */
@TargetApi(11)
public final class XpNumberPicker {
    private static final Field FIELD_SELECTION_DIVIDER;

    static {
        Field f = null;
        if (Build.VERSION.SDK_INT >= 11) {
            try {
                f = NumberPicker.class.getDeclaredField("mSelectionDivider");
                f.setAccessible(true);
            } catch (NoSuchFieldException e) {
            }
        }
        FIELD_SELECTION_DIVIDER = f;
    }

    private XpNumberPicker() {}

    public static void setSelectionDivider(NumberPicker picker, Drawable divider) {
        if (Build.VERSION.SDK_INT < 11) return;

        Drawable old = getSelectionDivider(picker);
        if (old != divider) {
            if (old != null) {
                old.setCallback(null);
            }
            try {
                FIELD_SELECTION_DIVIDER.set(picker, divider);
                if (divider != null) {
                    divider.setCallback(picker);
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        divider.setLayoutDirection(picker.getLayoutDirection());
                    }
                    if (divider.isStateful()) {
                        divider.setState(picker.getDrawableState());
                    }
                }
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        }
    }

    public static Drawable getSelectionDivider(NumberPicker picker) {
        if (Build.VERSION.SDK_INT < 11) return null;

        try {
            return (Drawable) FIELD_SELECTION_DIVIDER.get(picker);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static void setSelectionDividerTint(NumberPicker picker, ColorStateList color) {
        if (Build.VERSION.SDK_INT < 11) return;

        setSelectionDividerTintInt(picker, color);
    }

    static void setSelectionDividerTintInt(final NumberPicker picker, final ColorStateList color) {
        Drawable d = getSelectionDivider(picker);
        if (d != null) {
            d = DrawableCompat.wrap(d);
            DrawableCompat.setTintList(d, color);
            setSelectionDivider(picker, d);
        }
    }
}

File: picker/src/main/java/net/xpece/android/picker/widget/XpNumberPicker.kt
@file:JvmName("XpNumberPicker")

package net.xpece.android.picker.widget

import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.Log
import android.widget.NumberPicker
import androidx.annotation.ColorInt
import androidx.core.graphics.drawable.DrawableCompat
import java.lang.reflect.Field

private object NumberPickerReflection {

    val FIELD_SELECTION_DIVIDER: Field =
            NumberPicker::class.java
                    .getDeclaredField("mSelectionDivider")
                    .apply { isAccessible = true }
}

/**
 * Used to change [NumberPicker] selection divider drawable on-the-fly.
 *
 * **WARNING!** This method doesn't work on Android 10 and newer. There is no alternative.
 *
 * @see [setSelectionDividerTintCompat]
 */
@Suppress("DEPRECATION")
@Deprecated("Not available on API 29+.")
var NumberPicker.selectionDividerCompat: Drawable?
    get() = trySilent { NumberPickerReflection.FIELD_SELECTION_DIVIDER.get(this) as Drawable? }
    set(divider) {
        trySilent {
            val old = selectionDividerCompat
            if (old !== divider) {
                NumberPickerReflection.FIELD_SELECTION_DIVIDER.set(this, divider)

                old?.callback = null

                if (divider != null) {
                    divider.callback = this
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        divider.layoutDirection = layoutDirection
                    }
                    if (divider.isStateful) {
                        divider.state = drawableState
                    }
                }
            }
        }
    }

/**
 * Used to change [NumberPicker] selection divider color on-the-fly.
 *
 * This is primarily used on Android 4.x with AppCompat, so that the selection divider color
 * is `colorControlNormal` instead of Holo blue.
 *
 * You should rely on `android:colorControlNormal` on Android 5 and newer.
 *
 * **WARNING!** This method doesn't work on Android 10 and newer. There is no alternative.
 */
@Suppress("DEPRECATION")
fun NumberPicker.setSelectionDividerTintCompat(color: ColorStateList?) {
    var d = selectionDividerCompat
    if (d != null) {
        d = DrawableCompat.wrap(d)
        DrawableCompat.setTintList(d, color)
        selectionDividerCompat = d
    }
}

/**
 * Used to change [NumberPicker] selection divider color on-the-fly.
 *
 * This is primarily used on Android 4.x with AppCompat, so that the selection divider color
 * is `colorControlNormal` instead of Holo blue.
 *
 * You should rely on `android:colorControlNormal` on Android 5 and newer.
 *
 * **WARNING!** This method doesn't work on Android 10 and newer. There is no alternative.
 */
@Suppress("DEPRECATION")
fun NumberPicker.setSelectionDividerTintCompat(@ColorInt color: Int) {
    var d = selectionDividerCompat
    if (d != null) {
        d = DrawableCompat.wrap(d)
        DrawableCompat.setTint(d, color)
        selectionDividerCompat = d
    }
}

private inline fun <T> trySilent(block: () -> T): T? {
    if (Build.VERSION.SDK_INT < 29) {
        try {
            return block()
        } catch (ex: Throwable) {
            Log.e("XpNumberPicker", "Failed to access NumberPicker selection divider.", ex)
        }
    }
    return null
}
